### DISCUSSION TITLE
Image processing by example

### DISCUSSION THREAD

**Discussion by ksadil** [ImageMagick/ImageMagick#3609] [Digital Image Processing]
Hi, 

I am interested to know if I can use imagemagick to analyse two images as input, and determine what effects, color changes,transformations, etc it takes on the first, to make it look like the second. 

For context I have a very efficient process for taking photos on a dslr camera, wirelessly uploading to my phone and editing/processing on phone but this produces a low resolution image. So I want to use that processed-on-phone image as a specification and see what changes are required to transform the original image, then apply those changes to the high resolution photo in order to produce images that can be printed on large formats.

Given the brilliant way imagemagick is designed as a command line tool with image comparison features, it should be a relatively simple mathematical optimisation problem. But I do not want to reinvent the wheel, so to speak. 

Samples below of processed an unprocessed images:

![DSC04536](https://user-images.githubusercontent.com/24473342/116448538-95694200-a89c-11eb-9c50-132216f4fd00.JPG)

![DSC04536unprocessed](https://user-images.githubusercontent.com/24473342/116448715-c5184a00-a89c-11eb-9490-01d2ed191543.JPG)

**Comment by fmw42** [comment]
What is your Imagemagick version and platform/OS?  For Unix-like systems, I have a script that tries to match mean and standard deviation of one image to that of another. See matchimage at http://www.fmwconcepts.com/imagemagick.  For Windows, I believe that user snibgo has a similar .bat script at http://im.snibgo.com

**Reply by ksadil** [reply]
Great thanks. I use any system that achieves the results. I think I installed slackware linux in 1994 😁

**Comment by snibgo** [comment]
Do you start from raw camera images, or OOC (out-of-camera) JPEGs?

Do you make changes only globally, or do you also make local adjustments (such as darkening one corner)? Do any of your changes make results from multiple pixels (such as sharpening or blurring)?

Your examples suggest that your edits include cropping. Is that so? This makes analysis harder.

In the simplest case, you could set the means to be equal, or the means and contrast (using gain and bias, or power and sigmoidal-contrast), or adjust RGB channels with a 3x3 matrix, or match histograms, or use sparse Hald cluts.

**Reply by ksadil** [reply]
Wireless transferred jpegs are automatically downscaled. I do crop, rotate then do global changes. I suspect some tools in google photos use ai? Eg "pop"

Not expecting someone has done exactly what I want, but it would not surprise me. 

Not easy to have an original idea on the internet anymore.

**Comment by snibgo** [comment]
Ah, yes, I see your example includes a rotation as well as a crop. So, the first step would be to establish what rotation and crop was used (quite simple), so the colour analysis would use the exact same portions of the scene.

"Pop" is harder. That might include increasing _local_ contrast of lightness and colour saturation. Automatic analysis of that, and reproducing the effect, would not be trivial.

As you use a DSLR, I suggest you download and try image editors designed for the purpose, eg darktable or RawTherapee.

**Reply by ksadil** [reply]
I'm going only taking 5 minutes to edit, not a perfectionist. Even getting close would be very helpful.  And the more I think about it the more I like the idea
